sympy/physics/control/tests/test_lti.py

```diff
@@ -0,0 +1,268 @@
+    assert -tf3 == TransferFunction(3*p - 3, 1 - p, p)
+
+    # taking power of a TF.
+    tf4 = TransferFunction(p + 4, p - 3, p)
+    tf5 = TransferFunction(s**2 + 1, 1 - s, s)
+    expect2 = TransferFunction((s**2 + 1)**3, (1 - s)**3, s)
+    expect1 = TransferFunction((p + 4)**2, (p - 3)**2, p)
+    assert tf4*tf4 == tf4**2 == pow(tf4, 2) == expect1
+    assert tf5*tf5*tf5 == tf5**3 == pow(tf5, 3) == expect2
+    assert tf5**0 == pow(tf5, 0) == TransferFunction(1, 1, s)
+    assert tf4**-1 == pow(tf4, -1) == TransferFunction(p - 3, p + 4, p)
+    assert tf5**-2 == pow(tf5, -2) == TransferFunction((1 - s)**2, (s**2 + 1)**2, s)
+
+    raises(ValueError, lambda: tf4**(s**2 + s - 1))
+    raises(ValueError, lambda: tf5**s)
+    raises(ValueError, lambda: tf4**tf5)
+
+    # sympy's own functions.
+    assert tf3.xreplace({p: s}) == TransferFunction(-3*s + 3, 1 - s, s)
+    tf = TransferFunction(s - 1, s**2 - 2*s + 1, s)
+    tf6 = TransferFunction(s + p, p**2 - 5, s)
+    assert factor(tf) == TransferFunction(s - 1, (s - 1)**2, s)
+    assert tf.num.subs(s, 2) == tf.den.subs(s, 2) == 1
+    assert tf.subs(s, 2) == TransferFunction(s - 1, s**2 - 2*s + 1, s)
+    assert tf6.subs(p, 3) == TransferFunction(s + 3, 4, s)
+
+
+def test_TransferFunction_addition_and_subtraction():
+    tf1 = TransferFunction(s + 6, s - 5, s)
+    tf2 = TransferFunction(s + 3, s + 1, s)
+    tf3 = TransferFunction(s + 1, s**2 + s + 1, s)
+    tf4 = TransferFunction(p, 2 - p, p)
+
+    # addition
+    expect1 = TransferFunction((s - 5)*(s + 3) + (s + 1)*(s + 6), (s - 5)*(s + 1), s)
+    assert tf1 + tf2 == expect1
+    assert tf1 + (s - 1) == TransferFunction(s + (s - 5)*(s - 1) + 6, s - 5, s)
+    assert tf1 + 8 == TransferFunction(9*s - 34, s - 5, s)
+    assert (1 - p**3) + tf1 == TransferFunction(s + (1 - p**3)*(s - 5) + 6, s - 5, s)
+
+    c = symbols("c", commutative=False)
+    raises(ValueError, lambda: tf1 + Matrix([1, 2, 3]))
+    raises(ValueError, lambda: tf2 + c)
+    raises(ValueError, lambda: tf3 + tf4)
+
+    # subtraction
+    expect1 = TransferFunction((s + 1)*(s + 6) - (s - 5)*(s + 3), (s - 5)*(s + 1), s)
+    assert tf1 - tf2 == expect1
+    assert tf3 - tf2 == \
+        TransferFunction((s + 1)**2 - (s + 3)*(s**2 + s + 1), (s + 1)*(s**2 + s + 1), s)
+    assert tf1 - (s - 1) == TransferFunction(s - (s - 5)*(s - 1) + 6, s - 5, s)
+    assert tf1 - 8 == TransferFunction(-7*s + 46, s - 5, s)
+    assert (s + 5) - tf2 == TransferFunction(-s + (s + 1)*(s + 5) - 3, s + 1, s)
+    assert (1 + p**4) - tf1 == TransferFunction(-s + (p**4 + 1)*(s - 5) - 6, s - 5, s)
+
+    raises(ValueError, lambda: tf1 - Matrix([1, 2, 3]))
+    raises(ValueError, lambda: tf3 - tf4)
+
+    p1 = (s - 5)*(s + 1)**2 + ((s + 1)*(s + 6) - (s - 5)*(s + 3))*(s**2 + s + 1)
+    p2 = (s - 5)*(s + 1)*(s**2 + s + 1)
+    assert tf1 - tf2 + tf3 == TransferFunction(p1, p2, s)
+
+
+def test_TransferFunction_multiplication_and_division():
```

Comment:
I was thinking about this a bit and realize that we aren't following SymPy design patterns for operators. For example, don't we want multiplication to work like:

```
>>> H = TransferFunction(num, den, s)
>>> G = Function('G')(s)
>>> type(H*G)
Mul
>>> (H*G).args
(H, G)
```

Reply:
Also:

```
>>> (H*G).doit().args()
(G*H.num, H.den, s)
>>> type(H*G.doit())
TransferFunction
```

Reply:
@sylee957 and @oscarbenjamin Can one or more of you comment on this?

Reply:
If the symbolic transfer function algebra need to be defined, I think that a general superclass like `SystemModel` can be added and we can define `SystemModelSeriesConnect`, `SystemModelParallelConnect`, `SystemModelFeedbackConnect` which can also extend to interconnection with state space models or a matrix of transfer functions.

I'd reference how it's done in mathematica
https://reference.wolfram.com/language/ref/SystemsModelSeriesConnect.html
https://reference.wolfram.com/language/ref/SystemsModelFeedbackConnect.html

Reply:
Why would we want as "SeriesConnect" when `H*G` can work (for SISO TFs)? That's how we do it on paper and these objects are simply expressions. For descriptions of MIMO systems (collections of TFs or a state space description) you'll need these types of methods because you have to pair inputs and outputs, but I'd like SISO transfer functions to behave like mathematical expressions.

Reply:
One other note from my perspective is that the way mathematica has designed these transfer function tools isn't really appealing and it doesn't really map to good design patterns in sympy. We should use it for inspiration, but I'm not sure we should copy it.

Reply:
If it becomes `Mul`, it should inherently become the type of `Expr`,

But I wouldn't find much stuff from `Expr` useful for transfer functions. 
The reason we would use `Expr` is for computing derivative, expanding series, or apply functions like `sin, cos`.
But would they make sense for transfer functions?

Reply:
The derivative of the magnitude of a transfer function as a function of the complex value is an important value in control design, as one example.

But I just want to be able to do block diagram algebra with transfer function objects like we do on paper. For example, you should be able to do this:
![zBMx7](https://user-images.githubusercontent.com/276007/84522243-20dce300-ac8b-11ea-928a-72a3d3d4cd2c.png)

There are certainly things that you can do with a general expression in SymPy that we shouldn't be able to do on a transfer function, but it seems that the core functionality for algebra is needed.

Reply:
Here is another nice example:

https://www.youtube.com/watch?v=xTlobODwG5A

We should be able to type sympy code using transfer objects that follows that video almost line by line and end up with a transfer function object that we can then utlize the many special methods we can attach to it, like `.bode()`, `.freq_response()`, `.zero_frequency_gain()`, etc. etc.

Reply:
@moorepants it's not clear to me what you are imagining here.

Is it something like:
```
>>> tf1 = TransferFunction(...)
>>> tf2 = TransferFunction(...)
>>> tf1 * tf2
TransferFunction(...)
```
Or are you thinking of something like
```
>>> tf1 * tf2
Series(TransferFunction(...), TransferFunction(...))
```
where `Series` is another class that shares a similar interface as the TransferFunction?

It looks like Mathematica does the former which is easier to implement but doesn't give the sense of seeing how the transfer function has been algebraically composed in the way that you have described.

Reply:
The second example you show what I'm asking about. I think that a transfer function object should act like a sympy expression but that expression should be limited to the definition and properties of a transfer function and that object should have unique methods associated with a transfer function. Here is a session that uses your Series object idea to do the block diagram algebra for a simple negative feedback loop with proportional gain.

```python
>>>
>>> plant = TransferFunction(1, s**2 + 2*zeta*wn*s + wn**2, s)
>>> plant.root_locus(subs={zeta:0.1, wn: 1})  # plots a root locus for simple proportional feedback
>>> controller = TransferFunction(k, 1, s)
>>> input = Function('X_d')(s)
>>> output = Function('X')(s)
>>> plant*controller
Series(TransferFunction(k, 1, s), TransferFunction(1, s**2 + 2*zeta*wn*s + wn**2, s))
>>> closed_loop_eq = Eq((input - output)*controller*plant, output))
>>> closed_loop_eq
Eq(Series(Parallel(X_d, X), TransferFunction(k, 1, s), TransferFunction(1, s**2 + 2*zeta*wn*s + wn**2, s)), X)
>>> closed_loop = solve(closed_loop_eq, output/input)
>>> feedback(plant, controller)  # this should produce the same thing in closed_loop or more simply a single transfer function
>>> closed_loop.den  # gives the closed loop characteristic equation
Parallel(1, Series(controller*plant))
>>> closed_loop.var
s
>>> closed_loop
Series(controller, plant, 1/(Parallel(1, TransferFunction(k, 1, s), TransferFunction(1, s**2 + 2*zeta*wn*s + wn**2, s)))
>>> closed_loop.doit()  # reduces the Series & Parallel objects to single combined transfer function
TransferFunction()
>>> closed_loop.bode(subs={k: 5, zeta: 0.2, wn: 1})  # creates a bode plot
>>> closed_loop.poles()  # calculates the poles (if it can)
>>> closed_loop.time_domain()  # returns the ODE this TF represents
>>> closed_loop.step_input_response()  # returns the TF respresent a step response
>>> closed_loop.step_input_response().to_time_domain()  # returns the solution to the ODE in the time domain if it can
>>> closed_loop.step_response_plot(subs={})  # plots a step response for specific numerical values
>>> closed_loop.zero_frequency_gains()  # takes limit as s -> 0
```

Hopefully that gives an idea. One issue in the PR is that the TranferFunction object is being designed with little thought of how it will actually be used. I really think that a high level set of tests that are derived from textbook transfer function problems will elucidate how someone would actually want to use this class. What I show above, is what seems very useful to me as a control engineer and a person that teaches controls.

We want to be able to define specific transfer functions (although a TransferFunctionExpr object like MatrixExpr may also be useful, maybe in place of Function('G')(s)), do some block diagram algebra to obtain new transfer functions, and then once you have the transfer functions you want, easily access all the fun and useful control engineering tools.

Reply:
I have meant the same idea for SystemModelSeriesConnect and Series as you have described.

But I’m not sure that the youtube example is easy to implement even with symbolic interconnection operators.

Because it can have some limitations like you need to connect blocks inside the smallest feedback loop before moving to the next feedback loop.

Reply:
> I have meant the same idea for SystemModelSeriesConnect and Series as you have described.

Ok, good to know. I don't really understand the mathematica code enough yet. That's my fault.

Reply:
> Because it can have some limitations like you need to connect blocks inside the smallest feedback loop before moving to the next feedback loop.

This is a very common need in classical siso controls.

Reply:
Okay, I agree on to have symbolic interconnection operators.
It's more easier to check arithmetic is done right by looking at the symbolic operators. Although I'm still afraid that the printed results can look too huge if we can't do anything about that, as you have hidden them by ellipsis.

But we may need to convey these ideas clearly to @namannimmo10 about in implementation detail.

Reply:
> Although I'm still afraid that the printed results can look too huge if we can't do anything about that, as you have hidden them by ellipsis.

Can't this be the same as normal expressions? srepr will show the long form, but repr will display a mathematical looking expression.

Reply:
Would it make more sense to have the TransferFunction be an attribute of some kind of more symbol-like Block object?

The idea would be to create a Block for each block in the block diagram and give the block a name and associated transfer function. The block itself displays as its name but you can ask it for its transfer function:
```python
>>> plant = Block('plant', TransferFunction(...))
>>> plant
plant
>>> plant.transfer_function
TransferFunction(...)
>>> controller = Block('controller', TransferFunction(...))
>>> system = plant * controller
>>> system
Series(plant, controller)
>>> system.transfer_function
TransferFunction(...)
```

Reply:
I'm not sure I see what the benefit is? Is this to prevent doing block diagram algebra directly on the transfer function?

Reply:
It means that you can see the relationships from the system from a high-level when building it up. The transfer functions might be really complicated but the blocks as you build up show the way that the system is connected rather than the transfer function itself.

It also means that you can give identity to components that have different roles in the system but the same transfer function e.g.:
```python
>>> plant1 = Block('plant1', tf)
>>> plant2 = Block('plant2', tf)
>>> plant3 = Block('plant3', tf)
>>> system = (plant1 + plant2) * plant3
>>> system
Series(Parallel(plant1, plant2), plant3)
```
Now you can do
```python
>>> alternate_plant = Block('alternate_plant', tf_other)
>>> system.subs(plant2, alternate_plant)
Series(Parallel(plant1, alternate_plant), plant3)
```
It's possible to substitute for plant2 alone because it is distinct from plant1 and plant3 even though all have the same transfer function.

Giving identity to the different components makes it possible to distinguish them when e.g. plotting a time series. If you wanted to have a function for showing the block diagram then you would be able to label the different blocks.

It's not really clear to me where this is all supposed to end up though. I'm imagining something like a non-graphical version of simulink but maybe that's not it. Your mockup above makes it unclear to me why there is a transfer function object at all rather than just working with expressions that represent it. The methods like `root_locus` etc could just be functions that take an expression as argument e.g.:
```python
>>> plant = 1 / (s**2 + 2*zeta*wn*s + wn**2)
>>> root_locus(plant.subs({zeta:0.1, wn: 1}), s)  # plots a root locus for simple proportional feedback
```

Reply:
FWIW, this type of input/output structure is present in the [python-control](https://github.com/python-control/python-control) `InputOutputSystem` class.

Reply:
> Your mockup above makes it unclear to me why there is a transfer function object at all rather than just working with expressions that represent it.

I think we have to have a transfer function object in this package, unless we would not be able to further extend this package in the future. A lot of other functionality would require working with that object, not just working with normal sympy expressions.

Reply:
> Your mockup above makes it unclear to me why there is a transfer function object at all rather than just working with expressions that represent it. The methods like root_locus etc could just be functions that take an expression as argument e.g.:

This is always true for any objects, you can always write a set of functions that provide equivalent functionality. There are several reasons to create an object: 1) a transfer function is a a mathematical thing just like vectors, matrices, expressions, and sympy is designed around mathematical objects it seems natural to include transfer functions in this category, 2) there are data that all functions that operate on transfer functions share in common that would be useful to have as attributes instead of function arguments, 3) it is pythonic to utilize objects, 3) most other libraries think it is a good idea to use an object to represent a transfer function, 4) transfer functions seem to naturally fit into the expression tree structure that sympy is based on, it's really just a special type of function, etc.

Reply:
> It's possible to substitute for plant2 alone because it is distinct from plant1 and plant3 even though all have the same transfer function.

This is useful, but seems like an extra layer of complexity for something that is already available. For example:

```
>>> plant1 = Function('G')(s)
>>> plant2 = Function('H')(s)
>>> plant3 = Function('J')(s)
# or
>>> plant1 = TransferFunctionExpr('G', s)
>>> plant2 = TransferFunctionExpr('H', s)
>>> plant3 = TransferFunctionExpr('J', s)
>>> system = (plant1 + plant2) * plant3
>>> system
Series(Parallel(plant1, plant2), plant3)
>>> alternate_plant = Function('K')(s)
>>> system.subs(plant2, alternate_plant)
Series(Parallel(plant1, alternate_plant), plant3)
>>> system.subs(plant1, TransferFunction(1, s**2 + 2+ 1, s))
```

Reply:
I guess I also don't see why block diagram algebra is unique from normal algebra, other than maybe being more limited. We can create a `Series()` class but why isn't `Mul()` sufficient already? Same for `Parallel()` and `Add()`.

Reply:
The reason that block diagram algebra is unique is that there are many different types of block diagrams (transfer function, state space, nonlinear state space, ...) and you can't easily expect the result of addition or multiplication when different types of block diagrams are connected. 

For example, the series or parallel connection state space blocks are clearly different than scalar algebra or matrix algebra and I don't think that `Add` or `Mul` are going to be useful for them than having a formal addition or multiplication objects. And I already experience that `Add` and `Mul` are terrible to use for objects that are not from `Expr` or `Function` because it implicitly expects many private methods from it.

Reply:
> The reason that block diagram algebra is unique is that there are many different types of block diagrams (transfer function, state space, nonlinear state space, ...) and you can't easily expect the result of addition or multiplication when different types of block diagrams are connected.

A SISO transfer function doesn't need to connect with those things (at least now). The pupose of this PR is to get a fully functional SISO transfer function that would let you solve a large swath of problems in classical controls. We can introduce new things to connect SISO to MIMO systems later. I completely understand why Mul, Add, Pow won't work for a broader set of objects, but this is strictly about SISO transfer functions.

Reply:
> but this is strictly about SISO transfer functions.

If you want to begin with small like SISO transfer function, why would you not be satisfied with in-place addition or multiplication as done now.

And I also think that finding the actual transfer function from the problems like
https://github.com/sympy/sympy/pull/19390#discussion_r439506439 
can be done with what sympy already had, using `solve` with some algebraic equations.

So I'd also consider `TransferFunction` to be sufficient as a final wrapper and can let the user solve the system equation elsewhere, and it doesn't really have to solve anything on its own right now.

Reply:
Note that if we have a SISO transfer function that works fine using sympy expression functionality then we get this by default:

```
>>> tf1 = TransferFunction()
>>> tf2 = TransferFunction()
>>> tf3 = TransferFunction()
>>> tf4 = TransferFunction()
>>> # 2 inputs, 3 outputs
>>> mimo_plant = Matrix([[tf1, tf2, tf3], [tf4, tf5, tf6]])  # or we may want a custom TransferFunctionMatrix()
>>> closed_loop_inputs = Matrix([Xd, Yd, Zd])
>>> closed_loop_outputs = Matrix([X, Y, Z])
>>> controller = Matrix([[k1 , k2], [k3, k4], [k5, k6]])
>>> sysetm_eq = (closed_loop_inputs - closed_loop_outputs)*controller*mimo_plant - closed_loop_outputs
```

As long as the user ensures the dimensional is correct, this mimo black diagram algebra should work too.

Reply:
> We can create a `Series()` class but why isn't `Mul()` sufficient already? Same for `Parallel()` and `Add()`

If the TransferFunction can't be an ordinary Expr because it needs additional attributes/methods then I expect the same would apply to using Add/Mul vs Parallel/Series. This is analogous to why we have `MatMul` for matrices because regular `Mul` has no `shape` attribute, doesn't check the consistency of the matrix dimensions etc.

Reply:
> If you want to begin with small like SISO transfer function, why would you not be satisfied with in-place addition or multiplication as done now.

I'm potentially willing to go with that, but I think that sets us up for people trying to use transfer function expressions with a whole host of sympy functions that should work, but won't. The example I give is solve() not working.

> And I also think that finding the actual transfer function from the problems like
#19390 (comment) can be done with what sympy already had, using solve with some algebraic equations.

Sure, we can do everything we need to do with current sympy tools, but the point of this new module is to remove lots of boiler plate and sympy details, providing end users with a higher level abstraction so that working with transfer functions is more like it is with pencil & paper.

> So I'd also consider TransferFunction to be sufficient as a final wrapper and can let the user solve the system equation elsewhere, and it doesn't really have to solve anything on its own right now.

I personally want TransferFunction to have lots of smart methods so I don't have to solve things manually. For example, `TransferFunciton.feedback()` should create a new transfer function of a simple closed loop system because it is such a common task. But I should also be able to do that manually if needed.

Reply:
> If the TransferFunction can't be an ordinary Expr because it needs additional attributes/methods then I expect the same would apply to using Add/Mul vs Parallel/Series. This is analogous to why we have MatMul for matrices because regular Mul has no shape attribute, doesn't check the consistency of the matrix dimensions etc.

Yes, if we need Series() and Parallel() to act like transfer functions, then we need these classes. Here is a comparison:

```
>>> tf1 = TransferFunction(1, s, s)
>>> tf2 = TransferFunction(1, s**2, s)
>>> tf1*tf2
Mul(TransferFunction(1, s, s), TransferFunction(1, s**2, s))
>>> to_transfer_function(tf1*tf2)
TransferFunction(1, s**3, s)
>>> to_transfer_function(tf1*tf2).root_locus()
```
If it is a Mul then we'd need to somehow convert to a TransferFunction object to be able to use all the properties of a transfer function.

But if we have:

```
>>> tf1 = TransferFunction(1, s, s)
>>> tf2 = TransferFunction(1, s**2, s)
>>> tf1*tf2
Series(TransferFunction(1, s, s), TransferFunction(1, s**2, s))
>>>
>>> (ttf1*tf2).root_locus()
```

then that works nicely. So, I see clearly now why we need Series() and Parallel(), which is synonymous with MatMul and MatAdd.

Reply:
Is it possible that a block from the block diagram would be represented in some way that is not a TransferFunction?

Reply:
Yes, I think that is a broader need that we haven't put much thought into yet. If we get SISO transfer functions working nicely, this gives SymPy the power to solve that large set of problems for SISO classical controls. We could certainly add a "Block" object down the road that allows a variety of things to be put into a block diagram. I hesitate to do that right now, because it needs a lot of careful thinking about what can go in the blocks.

Reply:
I wonder if it makes sense conceptually to say that you have two transfer functions arranged in series rather than saying that you have two blocks in series each of which is defined in terms of a transfer function. There are equivalences e.g.
```
Series(block1, block2).transfer_function == block1.transfer_function * block2.transfer_function
Parallel(block1, block2).transfer_function == block1.transfer_function + block2.transfer_function
```
but I would carefully distinguish between the blocks which are in series/parallel and the transfer functions which are added or multiplied. If that's a distinction that will be wanted down the road then I think it's best to make that distinction now even if right now the only elementary block types are those defined in terms of transfer functions.

To me it would be natural to use different operators for the blocks like `&` and `|` rather than `*` and `+`. Or perhaps combining in series might most naturally be represented with something like `block1 >> block2`.

Reply:
> To me it would be natural to use different operators for the blocks like & and | rather than * and +. Or perhaps combining in series might most naturally be represented with something like block1 >> block2.

That'll take a lot of convincing for me. I still want to rip out the overloaded `&` and `^` from sympy.physics.mechanics. The `*` and `+` are what is used it textbooks for these operations, as TFs are functions.

Reply:
> The `*` and `+` are what is used it textbooks for these operations, as TFs are functions.

Transfer functions are functions but the blocks in the block diagram are not even if common cases can be defined using transfer functions. If that distinction is blurred in the early design then it will be difficult to disentangle later.

(This point is not about the operators although I think it applies to them as well)